### ISSUE TITLE
Add new augmentation - Binding holes, punch holes, clip/pin mark

### ISSUE THREAD

**Issue by kwcckw** [sparkfish/augraphy#62]
From rvl-cdip dataset, i can see there's a lot images with binding holes, punch holes , clip or pin mark. So I think we can add this similar effect to our repo too?

Examples:
![img_315](https://user-images.githubusercontent.com/52521979/130757543-045c9eec-d662-4a02-b9b1-a78010c69d6d.png)
![img_206158](https://user-images.githubusercontent.com/52521979/130757546-c73b718d-e3ed-42d6-9273-08e46f74b529.png)
![img_218356](https://user-images.githubusercontent.com/52521979/130757550-7ad546d1-14db-4d7a-ac1c-d5464a3202db.png)
![img_218584](https://user-images.githubusercontent.com/52521979/130757557-a25cabd3-d4aa-49d4-9f7d-aa14f33acf17.png)
![img_67](https://user-images.githubusercontent.com/52521979/130757900-0e5744f8-a8a1-4711-94cb-1115ed3a2988.png)

**Comment by shaheryar1** [comment]
I also noticed this in  resumes and reports of [tobacco3482jpg](https://www.kaggle.com/patrickaudriaz/tobacco3482jpg) dataset. I think we should add this , also its pretty straight forward.

**Comment by proofconstruction** [comment]
This is a great idea to add, but I think we may want to try to do it synthetically (creating the "holes" with numpy/cv2) rather than including images of existing holes in a folder, and copying them onto the document.

There are a few kinds of holes I've seen so far:
1. circular holes, like on the left side of the last image above,
2. rectangular holes, like on the image above that,
3. trapezoidal holes, like the ones at the bottom of the last image above.

Each of these has some kind of shadow effect around it, which would have to be accounted for.

I suggest we crop these images to the holes and read them in with cv2, then examine the resulting numpy array. We can then probably use simple transformations of these arrays to produce different sizes, shapes, alignments, etc. of the holes, and then layer those over a document image in the post phase.

**Comment by proofconstruction** [comment]
I just added [an augmentation](https://github.com/sparkfish/augraphy/blob/dev/augraphy/utilities/imageoverlay.py) for overlaying an image (like a paperclip) on documents.

Currently, it randomly chooses where to put the image, and sometimes (intentionally) leaves some of the overlaid image off the result, like if only part of the paperclip was visible at the edge of the scan.

We may want to add rotation to the overlaid image at some point, as well as the ability to specify where to put it.

**Comment by kwcckw** [comment]
> 
> 
> I just added [an augmentation](https://github.com/sparkfish/augraphy/blob/dev/augraphy/utilities/imageoverlay.py) for overlaying an image (like a paperclip) on documents.
> 
> Currently, it randomly chooses where to put the image, and sometimes (intentionally) leaves some of the overlaid image off the result, like if only part of the paperclip was visible at the edge of the scan.
> 
> We may want to add rotation to the overlaid image at some point, as well as the ability to specify where to put it.

Could you show some examples?

**Comment by proofconstruction** [comment]
It's not very pretty, but using these images:

![lowresfax](https://user-images.githubusercontent.com/74747193/132096400-6ceb0d20-d276-4fef-aff6-ba0230cd9929.png)
![paperclip](https://user-images.githubusercontent.com/74747193/132096404-4084df9f-fe46-4275-a506-2ffddda3b8df.png)

and this pipeline:

```python
import numpy as np
import random
import cv2

from augraphy import *

img = cv2.imread("lowresfax.png")

paperclip = cv2.imread("paperclip.png")

ink = []

paper = []

post = [ImageOverlay(paperclip, p=1)]

pipeline = AugraphyPipeline(ink,paper,post)

complete = pipeline.augment(img)
```

these were generated

![overlaid](https://user-images.githubusercontent.com/74747193/132096422-9044a4d2-c4db-4f91-9d4c-8791399c8f19.png)
![overlaid3](https://user-images.githubusercontent.com/74747193/132096456-62487ab6-24ff-4dba-80f1-3b250c17b3c1.png)
![overlaid2](https://user-images.githubusercontent.com/74747193/132096423-fa660f2d-3479-4f72-b0e0-1a7d283918b7.png)

**Comment by proofconstruction** [comment]
Obvious enhancements are removing background from the overlay and rotating/scaling the overlay, so the composite image looks natural. I just cropped some screenshots quickly to show this works.

**Comment by jboarman** [comment]
As a generic image overlay tool, this is a pretty good option to have in the tool belt, especially once images are cleaned up with transparent backgrounds, etc.

Eventually, I do agree with your initially thoughts on aiming to come up with an algorithmic approach for synthesizing these artifacts.  But this implementation provides a crawl-before-we-walk improvement that provides value until we can sort out the more complicated approaches down the road.

Adding some presets on locations within the document to place the random overlays may help as well, like LEFT, TOP, CORNER?

**Comment by kwcckw** [comment]
I see, for those pins and clip, it could be anywhere in the image, but for binding holes or punch holes, i think it should be more relevant if they are in either left or top position. It would be very rare (i haven't seen any from the dataset) to see them in right side or bottom side.

**Comment by proofconstruction** [comment]
The main issue I see is that hole punches, binding holes, etc. are usually multiple images regularly spaced; to produce the effect correctly, there would be `n+1` calls to `ImageOverlay`, for `n` binding holes, to generate the final foreground image that gets overlaid on the document with the `n+1`th call.

Choosing how many binder holes should be overlaid is a matter for the user. We could provide a utility class to generate the final overlay image for these, given an image to copy, some number of copies to make, and the location to place the copies at.

**Comment by proofconstruction** [comment]
The [latest commit](https://github.com/sparkfish/augraphy/commit/f8a904cadaa14ece8d7074b35de82dfb9103052e) adds `augraphy/utilities/overlaybuilder.py`, which contains a class for building just such an overlay image mentioned in the previous comment.

Using this code (just using a white page and black rectangular holes as an example):

```python
import cv2
import numpy as np

bg = np.ones((1500,1000,3))
fg = np.zeros((20,50,3))
ob = OverlayBuilder(fg, bg, 30, "left", 10)
overlay = ob.buildOverlay()
```

it's able to produce output like this (this will show up as just a column of 30 black rectangles on GitHub because the page background is fully white):
![overlay](https://user-images.githubusercontent.com/74747193/132362163-62c80cf2-75ab-4a85-abd5-e119ae154cb4.jpg)

In addition to the background and foreground images, we can pass in:
1. the number of copies of the foreground image (the binder hole, etc) to make,
2. a page edge from ["left", "right", "top", "bottom"],
3. an offset from the page edge (in this example, the "binder holes" are 10 pixels from the left edge

**Comment by jboarman** [comment]
Nice!  This is great.  I like that your demo shows that we can use static or algorithmically-generated images as overlays.

**Comment by proofconstruction** [comment]
This is now resolved with the latest PR #93 and combinations of `OverlayBuilder` and `ImageOverlay`.

**Comment by kwcckw** [comment]
> 
> 
> This is now resolved with the latest PR #93 and combinations of `OverlayBuilder` and `ImageOverlay`.

So I think it is better to have another augmentation on this too? We can have some predefined effect such as holes, clip or pin, and user can just select types of effect, number of repetition, and side. The augmentation should be using OverlayBuilder or ImageOverlay.

**Comment by proofconstruction** [comment]
We could do that to make it easier, but I don't really want to include more images in the repository.

What about putting the images of holes, clips, pins, etc. into Figshare and using the `FigshareDownloader` inside this new augmentation? Then the default can be to download the image files and use them to make an overlay, but the user can also pass their own images in.

**Comment by kwcckw** [comment]
> 
> 
> We could do that to make it easier, but I don't really want to include more images in the repository.
> 
> What about putting the images of holes, clips, pins, etc. into Figshare and using the `FigshareDownloader` inside this new augmentation? Then the default can be to download the image files and use them to make an overlay, but the user can also pass their own images in.

Right, actually I think we can do as below:

1. Check if user pass in any image, if no, then proceed to to step 2.
2. Try to download those predefined images from Figshare, if there's is no internet connection or download failed, proceed to step 3.
3. Create some simple holes, pins or clips effect and apply overlay augmentation. (i think it should be still manageable to create some simple black&white effect from scratch).

**Comment by proofconstruction** [comment]
I agree; do you want to work on this?

**Comment by kwcckw** [comment]
> 
> 
> I agree; do you want to work on this?

Yea, i can work on this next, but do you still need to update those overlay related code? Or it should be completed now?

**Comment by kwcckw** [comment]
By the way, what would be a better name for this augmentation? I can't think of a relevant name now.

**Comment by jboarman** [comment]
I can think of anything perfectly succinct.  Perhaps `Fasteners` or `BindingsAndFasteners` work well enough and could be extended to later include staples, staple holes, etc.

**Comment by kwcckw** [comment]
>  I can think of anything perfectly succinct. Perhaps Fasteners or BindingsAndFasteners work well enough and could be extended to later include staples, staple holes, etc.

Thanks, i will proceed with `BindingsAndFasteners` now, we may change it again later if there's a better name.

**Comment by proofconstruction** [comment]
Merged #95
